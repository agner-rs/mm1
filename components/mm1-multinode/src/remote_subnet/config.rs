use std::net::SocketAddr;

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case", tag = "protocol")]
pub enum RemoteSubnetConfig {
    Wip(ProtocolWip),
}

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct ProtocolWip {
    pub(crate) codec: String,
    pub(crate) link:  Link,
    pub(crate) serde: SerdeFormat,
}

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Link {
    pub(crate) bind: SocketAddr,
    pub(crate) peer: SocketAddr,
}

#[derive(
    Debug, Clone, Copy, serde::Deserialize, serde::Serialize, PartialEq, PartialOrd, Eq, Ord, Hash,
)]
#[serde(rename_all = "snake_case")]
pub(crate) enum SerdeFormat {
    #[cfg(feature = "format-json")]
    Json,
    #[cfg(feature = "format-bincode")]
    Bincode,
    #[cfg(feature = "format-rmp")]
    Rmp,
}
