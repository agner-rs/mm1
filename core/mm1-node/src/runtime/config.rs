use std::collections::{HashMap, HashSet};

use mm1_address::subnet::NetAddress;
use tokio::runtime::Runtime;

use crate::runtime::actor_key::ActorKey;

mod actor_config;
mod rt_config;

pub(crate) use actor_config::EffectiveActorConfig;

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Mm1Config {
    #[cfg_attr(feature = "serde", serde(default = "defaults::subnet_address"))]
    pub(crate) subnet: NetAddress,

    #[cfg_attr(feature = "serde", serde(default))]
    actor: actor_config::ActorConfigNode,

    #[cfg_attr(feature = "serde", serde(default))]
    runtime: rt_config::RtConfigs,
}

impl Mm1Config {
    pub(crate) fn actor_config(&self, actor_key: &ActorKey) -> impl EffectiveActorConfig + '_ {
        self.actor.select(actor_key)
    }

    pub(crate) fn build_runtimes(&self) -> std::io::Result<(Runtime, HashMap<String, Runtime>)> {
        self.runtime.build_runtimes()
    }

    pub(crate) fn validate(&self) -> Result<(), String> {
        let runtime_keys: HashSet<_> = self.runtime.runtime_keys().collect();
        self.actor.ensure_runtime_keys_are_valid(&runtime_keys)?;

        Ok(())
    }
}

impl Default for Mm1Config {
    fn default() -> Self {
        Self {
            subnet:  consts::LOCAL_SUBNET,
            actor:   Default::default(),
            runtime: Default::default(),
        }
    }
}

pub mod consts {
    use mm1_address::address::Address;
    use mm1_address::subnet::{NetAddress, NetMask};

    pub const LOCAL_SUBNET: NetAddress = NetAddress {
        address: Address::from_u64(0xFFFF0000_00000000),
        mask:    NetMask::M_16,
    };

    pub const DEFAULT_ACTOR_NETMASK: NetMask = NetMask::M_56;
    pub const DEFAULT_ACTOR_INBOX_SIZE: usize = 1024;
}

#[cfg(feature = "serde")]
mod defaults {
    use mm1_address::subnet::NetAddress;

    use super::consts;
    pub(super) const fn subnet_address() -> NetAddress {
        consts::LOCAL_SUBNET
    }
}
