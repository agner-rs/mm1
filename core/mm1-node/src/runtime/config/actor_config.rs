use std::collections::{HashMap, HashSet};

use mm1_address::subnet::NetMask;

use super::ActorKey;

pub(crate) trait EffectiveActorConfig {
    fn netmask(&self) -> NetMask;
    fn inbox_size(&self) -> usize;
    fn fork_inbox_size(&self) -> usize;
    fn runtime_key(&self) -> Option<&str>;
}

#[derive(Debug, Clone, Default)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub(crate) struct ActorConfigNode {
    #[cfg_attr(feature = "serde", serde(flatten, default))]
    config: ActorConfig<String>,

    #[cfg_attr(feature = "serde", serde(rename = "/", default))]
    sub: HashMap<String, Self>,
}

#[derive(Debug, Clone, Default)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub(crate) struct ActorConfig<S> {
    runtime:         Option<S>,
    netmask:         Option<NetMask>,
    inbox_size:      Option<usize>,
    fork_inbox_size: Option<usize>,
}

impl ActorConfigNode {
    pub(crate) fn select(&self, actor_key: &ActorKey) -> impl EffectiveActorConfig + '_ {
        let mut out = ActorConfig {
            runtime:         self.config.runtime.as_ref(),
            netmask:         self.config.netmask,
            inbox_size:      self.config.inbox_size,
            fork_inbox_size: self.config.fork_inbox_size,
        };

        let mut node = self;
        for p in actor_key.path() {
            let Some(n) = node.sub.get(p).or_else(|| node.sub.get("_")) else {
                break
            };
            out.runtime = n.config.runtime.as_ref().or(out.runtime);
            out.netmask = n.config.netmask.or(out.netmask);
            out.inbox_size = n.config.inbox_size.or(out.inbox_size);
            out.fork_inbox_size = n.config.fork_inbox_size.or(out.fork_inbox_size);
            node = n;
        }

        out
    }

    pub(crate) fn ensure_runtime_keys_are_valid(
        &self,
        available_keys: &HashSet<&str>,
    ) -> Result<(), String> {
        let mut used_keys = HashSet::<&str>::new();

        let mut q = vec![self];
        while let Some(n) = q.pop() {
            if let Some(k) = n.config.runtime.as_ref() {
                used_keys.insert(k);
            }
            q.extend(n.sub.values());
        }

        let invalid_keys = used_keys.difference(available_keys).collect::<Vec<_>>();
        if invalid_keys.is_empty() {
            Ok(())
        } else {
            Err(format!("invalid rt-keys: {:?}", invalid_keys))
        }
    }
}

impl EffectiveActorConfig for ActorConfig<&'_ String> {
    fn fork_inbox_size(&self) -> usize {
        self.fork_inbox_size
            .unwrap_or(defaults::DEFAULT_FORK_INBOX_SIZE)
    }

    fn netmask(&self) -> NetMask {
        self.netmask.unwrap_or(defaults::DEFAULT_ACTOR_NETMASK)
    }

    fn inbox_size(&self) -> usize {
        self.inbox_size
            .unwrap_or(defaults::DEFAULT_ACTOR_INBOX_SIZE)
    }

    fn runtime_key(&self) -> Option<&str> {
        self.runtime.as_ref().map(|s| s.as_ref())
    }
}

mod defaults {
    use super::*;

    pub(super) const DEFAULT_ACTOR_NETMASK: NetMask = NetMask::M_56;
    pub(super) const DEFAULT_ACTOR_INBOX_SIZE: usize = 1024;
    pub(super) const DEFAULT_FORK_INBOX_SIZE: usize = 32;
}
