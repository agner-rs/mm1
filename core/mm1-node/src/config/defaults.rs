use mm1_address::address::Address;
use mm1_address::subnet::{NetAddress, NetMask};

use crate::config::{DeclareSubnet, Mm1NodeConfig, SubnetKind};

const DEFAULT_LOCAL_SUBNET_ADDRESS: NetAddress = NetAddress {
    address: Address::from_u64(0xFFFF0000_00000000),
    mask:    NetMask::M_16,
};

impl Default for Mm1NodeConfig {
    fn default() -> Self {
        Self {
            subnets: default_subnets(),
            codecs:  Default::default(),
            actor:   Default::default(),
            runtime: Default::default(),
        }
    }
}

pub(super) fn default_subnets() -> Vec<DeclareSubnet> {
    vec![DeclareSubnet {
        net_address: DEFAULT_LOCAL_SUBNET_ADDRESS,
        kind:        SubnetKind::Local,
    }]
}

#[cfg(test)]
mod tests {
    use crate::config::Mm1NodeConfig;

    #[test]
    fn default_mm1_node_config_is_valid() {
        Mm1NodeConfig::default()
            .validate()
            .expect("default config is not valid");
    }
}
