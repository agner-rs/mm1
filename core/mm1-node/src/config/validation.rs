use std::collections::BTreeSet;
use std::ops::Deref;

use mm1_address::address_range::AddressRange;
use mm1_address::subnet::NetAddress;
use scc::HashSet;

use crate::config::{DeclareSubnet, Mm1NodeConfig, SubnetKind};

#[derive(Debug, Clone)]
pub struct Valid<C>(C);

#[derive(Debug, thiserror::Error)]
#[error("validation error:\n{}", errors.iter().map(ToString::to_string).collect::<Vec<_>>().join("\n-"))]
pub struct ValidationError<T> {
    pub errors:   Vec<String>,
    pub rejected: Box<T>,
}

impl Mm1NodeConfig {
    pub fn validate(self) -> Result<Valid<Self>, ValidationError<Self>> {
        TryFrom::try_from(self)
    }
}

impl TryFrom<Mm1NodeConfig> for Valid<Mm1NodeConfig> {
    type Error = ValidationError<Mm1NodeConfig>;

    fn try_from(node_config: Mm1NodeConfig) -> Result<Self, Self::Error> {
        let errors = {
            let known_codecs = node_config.codecs.iter().map(|c| c.name.as_str()).collect();

            let runtime_keys = node_config.runtime.runtime_keys().collect();
            validate_subnets(&node_config.subnets, &known_codecs)
                .chain(
                    node_config
                        .actor
                        .ensure_runtime_keys_are_valid(&runtime_keys)
                        .err(),
                )
                .collect::<Vec<_>>()
        };

        if errors.is_empty() {
            Ok(Self(node_config))
        } else {
            Err(ValidationError {
                errors,
                rejected: Box::new(node_config),
            })
        }
    }
}

fn validate_subnets<'a>(
    subnets: impl IntoIterator<Item = &'a DeclareSubnet>,
    #[allow(unused_variables)] known_codecs: &HashSet<&str>,
) -> impl Iterator<Item = String> {
    let mut errors: Vec<String> = Default::default();
    let mut local_subnet: Option<NetAddress> = Default::default();
    let mut subnets_tree: BTreeSet<AddressRange> = Default::default();

    for subnet in subnets {
        let net_address = subnet.net_address;
        let address_range = AddressRange::from(net_address);
        if let Some(existing) = subnets_tree.get(&address_range).copied() {
            errors.push(format!(
                "conflicting subnets: {} vs {}",
                NetAddress::from(existing),
                net_address
            ));
        } else {
            subnets_tree.insert(address_range);
        }

        match (local_subnet, &subnet.kind) {
            (None, SubnetKind::Local) => {
                local_subnet = Some(subnet.net_address);
            },
            (Some(existing), SubnetKind::Local) => {
                errors.push(format!(
                    "a subnet of type 'local' defined more than once: at least {existing} and \
                     {net_address}"
                ));
            },
            #[cfg(feature = "multinode")]
            (_, SubnetKind::Remote(remote)) => {
                if !known_codecs.contains(remote.codec.as_str()) {
                    errors.push(format!(
                        "subnet {} referes unknown codec {:?}",
                        net_address, remote.codec
                    ));
                }
            },
        }
    }

    if local_subnet.is_none() {
        errors.push("exactly one subnet of type 'local' must be defined".into());
    }

    errors.into_iter()
}

impl<C> Deref for Valid<C> {
    type Target = C;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

#[cfg(test)]
mod tests {
    use crate::config::validation::Valid;

    #[test]
    fn ergonomics() {
        struct Config {
            field: u32,
        }
        impl Config {
            fn method(&self) -> u32 {
                self.field
            }
        }

        let valid_value = Valid(Config { field: 42 });
        assert_eq!(valid_value.field, 42);
        assert_eq!(valid_value.method(), 42);

        let valid_reference = Valid(&Config { field: 13 });
        assert_eq!(valid_reference.field, 13);
        assert_eq!(valid_reference.method(), 13);
    }
}
