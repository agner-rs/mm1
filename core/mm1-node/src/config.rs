use std::collections::HashMap;

use mm1_address::subnet::NetAddress;

mod actor_config;
mod rt_config;
mod validation;

pub(crate) use actor_config::EffectiveActorConfig;
use tokio::runtime::Runtime;
pub use validation::{Valid, ValidationError};

use crate::actor_key::ActorKey;

#[derive(Debug, Clone, serde::Deserialize)]
pub struct Mm1NodeConfig {
    #[serde(default = "defaults::default_subnets")]
    subnets: Vec<DeclareSubnet>,

    #[serde(default = "Default::default")]
    codecs: Vec<DeclareCodec>,

    #[serde(default)]
    actor: actor_config::ActorConfigNode,

    #[serde(default)]
    runtime: rt_config::RtConfigs,
}

impl Valid<Mm1NodeConfig> {
    pub(crate) fn actor_config(&self, actor_key: &ActorKey) -> impl EffectiveActorConfig + '_ {
        self.actor.select(actor_key)
    }

    pub(crate) fn build_runtimes(&self) -> std::io::Result<(Runtime, HashMap<String, Runtime>)> {
        self.runtime.build_runtimes()
    }

    pub(crate) fn local_subnet_address(&self) -> NetAddress {
        self.subnets
            .iter()
            .find_map(|s| matches!(s.kind, SubnetKind::Local).then_some(s.net_address))
            .expect("Valid<Mm1NodeConfig>")
    }

    #[cfg(feature = "multinode")]
    pub(crate) fn subnets(&self) -> impl Iterator<Item = &'_ DeclareSubnet> + use<'_> {
        self.subnets.iter()
    }
}

#[derive(Debug, Clone, serde::Deserialize)]
pub(crate) struct DeclareSubnet {
    pub(crate) net_address: NetAddress,

    #[serde(flatten)]
    pub(crate) kind: SubnetKind,
}

#[derive(Debug, Clone, serde::Deserialize)]
#[serde(rename_all = "snake_case", tag = "type")]
pub(crate) enum SubnetKind {
    Local,

    #[cfg(feature = "multinode")]
    Remote(RemoteSubnetConfig),
}

#[cfg(feature = "multinode")]
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) struct RemoteSubnetConfig {
    pub(crate) codec: String,

    #[serde(flatten)]
    opaque: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub(crate) struct DeclareCodec {
    pub(crate) name: String,
}

mod defaults;

#[cfg(test)]
mod test_serde;
#[cfg(test)]
mod test_validation;
