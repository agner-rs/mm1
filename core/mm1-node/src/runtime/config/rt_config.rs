use std::collections::HashMap;
use std::time::Duration;

use tokio::runtime::{Builder, Runtime};

#[derive(Debug, Clone, Default)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub(crate) struct RtConfigs {
    default: Option<RtConfig>,

    #[cfg_attr(feature = "serde", serde(flatten))]
    named: HashMap<String, RtConfig>,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
struct RtConfig {
    #[cfg_attr(feature = "serde", serde(default = "defaults::worker_threads"))]
    worker_threads:             usize,
    #[cfg_attr(feature = "serde", serde(default = "defaults::blocking_threads_max"))]
    blocking_threads_max:       usize,
    #[cfg_attr(
        feature = "serde",
        serde(default = "defaults::blocking_thread_keep_alive")
    )]
    blocking_thread_keep_alive: Duration,
    #[cfg_attr(feature = "serde", serde(default = "defaults::thread_stack_size"))]
    thread_stack_size:          usize,
    #[cfg_attr(feature = "serde", serde(default = "defaults::enable_io"))]
    enable_io:                  bool,
    #[cfg_attr(feature = "serde", serde(default = "defaults::enable_time"))]
    enable_time:                bool,
}

impl RtConfigs {
    pub(crate) fn build_runtimes(&self) -> std::io::Result<(Runtime, HashMap<String, Runtime>)> {
        let default = if let Some(c) = self.default.as_ref() {
            c.build_runtime("default")?
        } else {
            Builder::new_multi_thread()
                .thread_name("default")
                .enable_all()
                .build()?
        };

        let mut named = HashMap::new();

        for (n, c) in &self.named {
            named.insert(n.to_owned(), c.build_runtime(n)?);
        }

        Ok((default, named))
    }

    pub(crate) fn runtime_keys(&self) -> impl Iterator<Item = &str> {
        self.named.keys().map(String::as_str)
    }
}

impl RtConfig {
    fn build_runtime(&self, thread_name: &str) -> std::io::Result<Runtime> {
        let mut b = Builder::new_multi_thread();
        b.thread_name(thread_name)
            .worker_threads(self.worker_threads)
            .max_blocking_threads(self.blocking_threads_max)
            .thread_keep_alive(self.blocking_thread_keep_alive)
            .thread_stack_size(self.thread_stack_size);

        if self.enable_io {
            b.enable_io();
        }
        if self.enable_time {
            b.enable_time();
        }

        let runtime = b.build()?;
        Ok(runtime)
    }
}

mod defaults {
    use std::time::Duration;

    use super::RtConfig;

    pub(super) fn worker_threads() -> usize {
        num_cpus::get()
    }
    pub(super) fn blocking_threads_max() -> usize {
        // taken from tokio's docs
        512
    }
    pub(super) fn blocking_thread_keep_alive() -> Duration {
        Duration::from_secs(10)
    }
    pub(super) fn thread_stack_size() -> usize {
        2 * 1024 * 1024
    }
    pub(super) fn enable_io() -> bool {
        true
    }
    pub(super) fn enable_time() -> bool {
        true
    }

    impl Default for RtConfig {
        fn default() -> Self {
            Self {
                worker_threads:             worker_threads(),
                blocking_threads_max:       blocking_threads_max(),
                blocking_thread_keep_alive: blocking_thread_keep_alive(),
                thread_stack_size:          thread_stack_size(),
                enable_io:                  enable_io(),
                enable_time:                enable_time(),
            }
        }
    }
}
