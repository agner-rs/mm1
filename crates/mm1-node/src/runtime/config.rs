use std::collections::{HashMap, HashSet};

use mm1_address::subnet::NetAddress;
use tokio::runtime::Runtime;

use crate::runtime::actor_key::ActorKey;

mod actor_config;
mod rt_config;

pub(crate) use actor_config::EffectiveActorConfig;

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Mm1Config {
    #[serde(default = "defaults::subnet_address")]
    pub(crate) subnet_address: NetAddress,

    #[serde(default)]
    actor: actor_config::ActorConfigNode,

    #[serde(default)]
    runtime: rt_config::RtConfigs,
}

impl Mm1Config {
    pub(crate) fn actor_config(&self, actor_key: &ActorKey) -> impl EffectiveActorConfig + '_ {
        self.actor.select(actor_key)
    }

    pub(crate) fn build_runtimes(&self) -> std::io::Result<(Runtime, HashMap<String, Runtime>)> {
        self.runtime.build_runtimes()
    }

    pub(crate) fn validate(&self) -> Result<(), String> {
        let runtime_keys: HashSet<_> = self.runtime.runtime_keys().collect();
        self.actor.ensure_runtime_keys_are_valid(&runtime_keys)?;

        Ok(())
    }
}

impl Default for Mm1Config {
    fn default() -> Self {
        Self {
            subnet_address: consts::DEFAULT_SUBNET_ADDRESS,
            actor:          Default::default(),
            runtime:        Default::default(),
        }
    }
}

pub mod consts {
    use mm1_address::address::Address;
    use mm1_address::subnet::{NetAddress, NetMask};

    pub const DEFAULT_SUBNET_ADDRESS: NetAddress = NetAddress {
        address: Address::from_u64(0xFFFFFFFF_00000000),
        mask:    NetMask::M_32,
    };

    pub const DEFAULT_ACTOR_NETMASK: NetMask = NetMask::M_56;
    pub const DEFAULT_ACTOR_INBOX_SIZE: usize = 1024;
}

mod defaults {
    use mm1_address::subnet::NetAddress;

    use super::consts;
    pub(super) const fn subnet_address() -> NetAddress {
        consts::DEFAULT_SUBNET_ADDRESS
    }
}
