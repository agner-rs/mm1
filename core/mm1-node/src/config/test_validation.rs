#![cfg(feature = "multinode")]

use std::collections::HashSet;

use insta::assert_yaml_snapshot;
use test_case::test_case;

use crate::config::*;

#[test_case(
    r#"
    "#
    ; "the default config is valid"
)]
#[test_case(
    r#"
    subnets:
        - net_address: <cafe:>/16
          type: local
    "#
    ; "single local subnet"
)]
#[test_case(
    r#"
    subnets:
        - net_address: <cafe:>/16
          type: local
        - net_address: <beef:>/16
          type: local
    "#
    ; "duplicate local subnet"
)]
#[test_case(
    r#"
    codecs:
        - name: some-codec
    subnets:
        - net_address: <cafe:>/16
          type: local
        - net_address: <beef:>/16
          type: remote
          codec: some-codec
    "#
    ; "two distinct subnets, one of them â€” local"
)]
#[test_case(
    r#"
    codecs:
        - name: some-codec
    subnets:
        - net_address: <cafe:>/16
          type: local
        - net_address: <cafe:>/16
          type: remote
          codec: some-codec
    "#
    ; "two subnets, same net-address"
)]
#[test_case(
    r#"
    codecs:
        - name: some-codec
    subnets:
        - net_address: <cafe:>/16
          type: local
        - net_address: <cafeffff:>/32
          type: remote
          codec: some-codec
    "#
    ; "two subnets, overlapping net-address"
)]
fn test_validation(yaml: &str) {
    let actual_errors = serde_yaml::from_str::<Mm1NodeConfig>(yaml)
        .expect("invalid format")
        .validate()
        .err()
        .into_iter()
        .flat_map(|e| e.errors)
        .collect::<HashSet<_>>();

    let thread_name = std::thread::current()
        .name()
        .map(ToOwned::to_owned)
        .unwrap();
    assert_yaml_snapshot!(thread_name, actual_errors);
}
