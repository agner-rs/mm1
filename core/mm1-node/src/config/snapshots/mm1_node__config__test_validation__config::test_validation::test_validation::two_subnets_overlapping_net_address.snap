---
source: core/mm1-node/src/config/test_validation.rs
expression: actual_errors
---
- "conflicting subnets: <cafe:>/16 vs <cafeffff:>/32"
